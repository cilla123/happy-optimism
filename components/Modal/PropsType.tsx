export interface BaseModalProps {
  shape?: 'radius';
  visible?: boolean;
  animationType?:
    'fade' | 'door' | 'flip' | 'rotate' | 'zoom' |
    'moveUp' | 'moveDown' | 'moveLeft' | 'moveRight' |
    'slideUp' | 'slideDown' | 'slideLeft' | 'slideRight';
  animationDuration?: number;
  width?: string | number;
  onMaskClick?: () => void;
}

export interface BaseModalHeaderProps {
  title?: any;
  onClose?: () => void;
}

export interface BaseModalBodyProps {
  height?: string | number;
}

export interface BaseModalFooterProps {
  block?: boolean;
}
