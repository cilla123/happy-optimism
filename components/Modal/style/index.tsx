import '../../style/';
import '../../Mask/style/';
import '../../Icon/style/';
import './index.scss';
