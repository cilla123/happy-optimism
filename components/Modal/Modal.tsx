import React, { PureComponent, CSSProperties } from 'react';
import classnames from 'classnames';
import { BaseModalProps } from './PropsType';
import Events from '../utils/events';
import Mask from '../Mask';

const stopPropagation = (e) => {
  e.stopPropagation();
  // e.nativeEvent.stopImmediatePropagation();
};

export interface ModalProps extends BaseModalProps {
  prefixCls?: string;
  className?: string;
}

export default class Modal extends PureComponent<ModalProps, any> {

  static Header: any;
  static Body: any;
  static Footer: any;
  static defaultProps = {
    prefixCls: 'za-modal',
    visible: false,
    animationType: 'fade',
    animationDuration: 200,
    width: '70%',
  };

  private modal;

  constructor(props) {
    super(props);
    this.state = {
      isShow: props.visible || false,
      isPending: false,
      animationState: 'enter',
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!this.props.visible && nextProps.visible) {
      this.enter();
    } else if (this.props.visible && !nextProps.visible) {
      this.leave();
    }
  }

  componentWillUpdate() {
    Events.on(this.modal, 'webkitAnimationEnd', this.animationEnd);
    Events.on(this.modal, 'animationend', this.animationEnd);
  }

  componentWillUnmount() {
    Events.off(this.modal, 'webkitAnimationEnd', this.animationEnd);
    Events.off(this.modal, 'animationend', this.animationEnd);
  }

  animationEnd = () => {
    if (this.state.animationState === 'leave') {
      this.setState({
        isShow: false,
        isPending: false,
      });
    } else {
      this.setState({
        isShow: true,
        isPending: false,
      });
    }
  }

  enter = () => {
    this.setState({
      isShow: true,
      isPending: true,
      animationState: 'enter',
    });
  }

  leave = () => {
    this.setState({
      isShow: true,
      isPending: true,
      animationState: 'leave',
    });
  }

  render() {
    const { prefixCls, className, shape, animationType, animationDuration, width, onMaskClick, children } = this.props;
    const { isShow, isPending, animationState } = this.state;

    const cls = {
      modal: classnames(`${prefixCls}`, className, {
        [`shape-${shape}`]: !!shape,
        [`fade-${animationState}`]: isPending,
      }),
      dialog: classnames(`${prefixCls}-dialog`, {
        [`${animationType}-${animationState}`]: isPending,
        [`fade-${animationState}`]: isPending,
      }),
      // mask: classnames({
      //   [`fade-${animationState}`]: isPending,
      // }),
    };

    const modalStyle: CSSProperties = {
      WebkitAnimationDuration: `${animationDuration}ms`,
      animationDuration: `${animationDuration}ms`,
    };

    const dialogStyle: CSSProperties = {
      width,
      WebkitAnimationDuration: `${animationDuration}ms`,
      animationDuration: `${animationDuration}ms`,
    };

    // const maskStyle: CSSProperties = {
    //   WebkitAnimationDuration: `${animationDuration}ms`,
    //   MozAnimationDuration: `${animationDuration}ms`,
    //   msAnimationDuration: `${animationDuration}ms`,
    //   OAnimationDuration: `${animationDuration}ms`,
    //   animationDuration: `${animationDuration}ms`,
    // };

    if (!isShow) {
      modalStyle.display = 'none';
    }

    return (
      <div className={cls.modal} style={modalStyle} ref={(ele) => { this.modal = ele; }}>
        <div className={`${prefixCls}-wrapper`}>
          <div className={cls.dialog} style={dialogStyle} onClick={stopPropagation}>
            {children}
          </div>
        </div>
        <Mask
          visible={isShow}
          // className={cls.mask}
          // style={maskStyle}
          onClose={onMaskClick}
        />
      </div>
    );
  }
}
