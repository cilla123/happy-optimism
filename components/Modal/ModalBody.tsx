import React, { PureComponent, CSSProperties } from 'react';
import classnames from 'classnames';
import { BaseModalBodyProps } from './PropsType';

export interface ModalBodyProps extends BaseModalBodyProps {
  prefixCls?: string;
  className?: string;
}

export default class ModalBody extends PureComponent<ModalBodyProps, {}> {

  static defaultProps = {
    prefixCls: 'za-modal',
  };

  render() {
    const { prefixCls, className, height, children, ...others } = this.props;
    const cls = classnames(`${prefixCls}-body`, className);

    const bodyStyle: CSSProperties = {};
    bodyStyle.height = height;

    return (
      <div className={cls} style={bodyStyle} {...others}>
        {children}
      </div>
    );
  }
}
