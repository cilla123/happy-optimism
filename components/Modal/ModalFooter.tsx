import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { BaseModalFooterProps } from './PropsType';

export interface ModalFooterProps extends BaseModalFooterProps {
  prefixCls?: string;
  className?: string;
}

export default class ModalFooter extends PureComponent<ModalFooterProps, {}> {

  static defaultProps = {
    prefixCls: 'za-modal',
  };

  render() {
    const { prefixCls, className, block, children, ...others } = this.props;
    const cls = classnames(`${prefixCls}-footer`, className, {
      block,
    });

    return (
      <div className={cls} {...others}>
        {children}
      </div>
    );
  }
}
