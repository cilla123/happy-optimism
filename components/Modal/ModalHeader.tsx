import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { BaseModalHeaderProps } from './PropsType';
import Icon from '../Icon';

export interface ModalHeaderProps extends BaseModalHeaderProps {
  prefixCls?: string;
  className?: string;
}

export default class ModalHeader extends PureComponent<ModalHeaderProps, {}> {

  static defaultProps = {
    prefixCls: 'za-modal',
    title: '',
  };

  render() {
    const { prefixCls, className, title, onClose, ...others } = this.props;
    const cls = classnames(`${prefixCls}-header`, className);
    const btnClose = onClose && <Icon type="wrong" className={`${prefixCls}-header-close`} onClick={onClose} />;

    return (
      <div className={cls} {...others}>
        <div className={`${prefixCls}-header-title`}>{title}</div>
        {btnClose}
      </div>
    );
  }
}
