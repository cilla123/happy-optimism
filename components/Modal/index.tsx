import Modal from './Modal';
import ModalHeader from './ModalHeader';
import ModalBody from './ModalBody';
import ModalFooter from './ModalFooter';

Modal.Header = ModalHeader;
Modal.Body = ModalBody;
Modal.Footer = ModalFooter;

export default Modal;
