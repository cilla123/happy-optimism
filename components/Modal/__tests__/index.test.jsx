import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Modal from '../index';

describe('Modal', () => {
  it('renders correctly', () => {
    const onMaskClick = jest.fn();
    const onClose = jest.fn();

    const wrapper = mount(
      <Modal visible onMaskClick={onMaskClick}>
        <Modal.Header title="标题" onClose={onClose} />
        <Modal.Body>foo</Modal.Body>
      </Modal>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('onClose', () => {
    const onClose = jest.fn();
    const wrapper = shallow(
      <Modal visible>
        <Modal.Header title="标题" onClose={onClose} />
        <Modal.Body>foo</Modal.Body>
      </Modal>
    );
    wrapper.find(Modal.Header).dive().find('.za-modal-header-close').simulate('click');
    expect(onClose).toBeCalled();
  });

  it('receive new visible', () => {
    const wrapper = mount(<Modal>foo</Modal>);
    jest.useFakeTimers();
    wrapper.setProps({ visible: true });
    jest.runAllTimers();
    wrapper.setProps({ visible: false });
    jest.runAllTimers();
  });

  it('click dialog', () => {
    const wrapper = mount(<Modal visible>foo</Modal>);
    wrapper.find('.za-modal-dialog').simulate('click');
  });
});
