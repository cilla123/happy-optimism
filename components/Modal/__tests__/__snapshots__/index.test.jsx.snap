// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<Modal
  animationDuration={200}
  animationType="fade"
  onMaskClick={[Function]}
  prefixCls="za-modal"
  visible={true}
  width="70%"
>
  <div
    className="za-modal"
    style={
      Object {
        "WebkitAnimationDuration": "200ms",
        "animationDuration": "200ms",
      }
    }
  >
    <div
      className="za-modal-wrapper"
    >
      <div
        className="za-modal-dialog"
        onClick={[Function]}
        style={
          Object {
            "WebkitAnimationDuration": "200ms",
            "animationDuration": "200ms",
            "width": "70%",
          }
        }
      >
        <ModalHeader
          onClose={[Function]}
          prefixCls="za-modal"
          title="标题"
        >
          <div
            className="za-modal-header"
          >
            <div
              className="za-modal-header-title"
            >
              标题
            </div>
            <Icon
              className="za-modal-header-close"
              onClick={[Function]}
              prefixCls="za-icon"
              type="wrong"
            >
              <i
                className="za-icon za-modal-header-close za-icon-wrong"
                onClick={[Function]}
              />
            </Icon>
          </div>
        </ModalHeader>
        <ModalBody
          prefixCls="za-modal"
        >
          <div
            className="za-modal-body"
            style={
              Object {
                "height": undefined,
              }
            }
          >
            foo
          </div>
        </ModalBody>
      </div>
    </div>
    <Mask
      onClose={[Function]}
      prefixCls="za-mask"
      type="normal"
      visible={true}
    >
      <div
        className="za-mask normal"
        onClick={[Function]}
      />
    </Mask>
  </div>
</Modal>
`;
